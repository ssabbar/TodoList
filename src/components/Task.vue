<template>
  <div class="mb-4">
    <div v-if="edited">
      <EditTask
      :task="task"
      @updateEditedValue="updateEditedValue"/>
    </div>
    <div v-else :id="'taskid' + task.id">
      <div class="mb-2">
        <input type="checkbox" @click="checkTask(task.id)" v-model="task.completed"/>
        <label :id="task.id" class="checkbox-label">{{task.name}}</label>
      </div>
      <div class="row">
        <button @click="updateEditedValue(true)" type="button" id="editTaskButton" class="btn btn-outline-secondary">Edit</button>
        <button @click="deleteTask(task.id)" type="button" id="deleteTaskButton" class="btn btn-danger">Delete</button>
      </div>
    </div>
  </div>
</template>

<script>
import EditTask from './EditTask.vue'
export default {
  name: 'Task',
  components: {EditTask},
  data () {
    return {
      edited: false
    }
  },
  props: {
    task: Object
  },
  methods: {
    checkTask (taskId) {
      this.$store.dispatch({
        type: 'checkTask',
        taskId
      })
    },
    updateEditedValue (value) {
      this.edited = value
    },
    deleteTask (taskId) {
      this.$store.dispatch({
        type: 'deleteTask',
        taskId: taskId
      })
    }
  }
}
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped>
input[type=checkbox] {
  transform: scale(2);
  margin: 0.5rem;
}

.checkbox-label {
  font-size: 115%;
  color: #0b0c0c;
}

button[type=button] {
  display: inline-block;
  width: 45%;
  margin-left: 1rem;
}

</style>
