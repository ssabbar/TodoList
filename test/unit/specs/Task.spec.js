import Vue from 'vue'
import Vuex from 'vuex'
import { mount, createLocalVue } from '@vue/test-utils'
import Task from '@/components/Task'

const localVue = createLocalVue()

localVue.use(Vuex)

describe('Task components', () => {
  let actions
  let store

  const mockedPropData = {
    task: {
      id: 1,
      name: 'task1',
      completed: false
    }
  }

  beforeEach(() => {
    actions = {
      checkTask: jest.fn(),
      deleteTask: jest.fn()
    }
    store = new Vuex.Store({
      actions
    })
  })

  it('renders correctly', async () => {
    const wrapper = mount(Task, { propsData: { ...mockedPropData } })
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setData({ edited: true })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('call store action "checkTask" when click on the checkbox of task1', async () => {
    const wrapper = mount(Task, {
      store,
      localVue,
      propsData: { ...mockedPropData }
    })

    await wrapper.find('input[type=checkbox]').trigger('click')
    expect(actions.checkTask).toHaveBeenCalledWith(expect.anything(), { type: 'checkTask', taskId: 1 })
  })

  it('call store action "deletekTask" when click on delete button of task1', async () => {
    const wrapper = mount(Task, {
      store,
      localVue,
      propsData: { ...mockedPropData }
    })

    await wrapper.find('#deleteTaskButton').trigger('click')
    expect(actions.deleteTask).toHaveBeenCalledWith(expect.anything(), { type: 'deleteTask', taskId: 1 })
  })

  it('call function updateEditedValue when click on edit button of task1', async () => {
    const wrapper = mount(Task, {
      store,
      localVue,
      propsData: { ...mockedPropData }
    })

    await wrapper.find('#editTaskButton').trigger('click')
    expect(wrapper.vm.edited).toBe(true)
  })
})
