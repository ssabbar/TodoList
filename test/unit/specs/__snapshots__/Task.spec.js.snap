// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task components renders correctly 1`] = `
<div class="mb-4">
    <div id="taskid1">
        <div class="mb-2"><input type="checkbox"> <label id="1" class="checkbox-label">task1</label></div>
        <div class="row"><button type="button" id="editTaskButton" class="btn btn-outline-secondary">Edit</button> <button type="button" id="deleteTaskButton" class="btn btn-danger">Delete</button></div>
    </div>
</div>
`;

exports[`Task components renders correctly 2`] = `
<div class="mb-4">
    <div>
        <div>
            <div>
                <p class="mb-2">Edit name for task1</p> <input class="mb-3">
            </div>
            <div class="row"><button id="cancelEditButton" class="btn btn-outline-secondary">Cancel</button> <button id="saveEditButton" class="btn btn-dark">Save</button></div>
        </div>
    </div>
</div>
`;
